### PR TITLE
Scc 2052 / add comparison of incoming data to existing headings in SHEP API

### PR DESCRIPTION
**Pull request by siberry**
This PR adds a `GET` request to SHEP API for the bib from the Bib stream record to check the differences in subject headings so that we only post to SHEP API if there is a change to subject headings.

The code for the `SHEP` class and corresponding tests is adapted from [subject-headings-explorer-poc](https://github.com/NYPL/subject-headings-explorer-poc/tree/shep-api).

### CODE REVIEW

File: app.rb [hunk 1]
```diff
@@ -28,24 +29,74 @@ def handle_event(event:, context:)
...
     end
 end
+
+def should_process data
+  return false unless is_research? data
+
+  return false unless have_subject_headings_changed? data
+
+  return true
+end
```
Comment (danamansana) [inline app.rb:51]: This could just be `is_research? (data) && have_subject_headings_changed? (data)`

File: app.rb [hunk 2]
```diff
@@ -28,24 +29,74 @@ def handle_event(event:, context:)
...
+  incoming_tagged_subject_headings = bib_data.heading_data_mgrs.map(&:tagged_label).sort
+
+  discovery_id = bib_data.discovery_id
+
+  # make GET request to SHEP API Bib#tagged_subject_headings endpoint
+  uri = URI("#{ENV['SHEP_API_BIBS_ENDPOINT']}#{discovery_id}/tagged_subject_headings")
+  resp = Net::HTTP.get_response(uri)
+
+  preexisting_tagged_subject_headings = JSON.parse(resp.body)["tagged_subject_headings"].sort if resp.code == "200"
+
```
Comment (danamansana) [inline app.rb:94]: What if `resp.code` isn't 200? It looks like `preexisting_tagged_subject_headings` is still required on line 94

File: lib/bib_data_manager.rb [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+        @publication_year = process_publication_year publication_year_fields
+
+        @heading_data_mgrs = headings_data(headings_fields).map do |heading_data|
+          next unless subject_heading_length_ok? heading_data[:subfields]
+
+          SubjectHeadingDataManager.new_from_data(heading_data)
+        end.compact
+      end
+
+      def extract_useful_var_fields(var_fields)
```
Comment (danamansana) [inline lib/bib_data_manager.rb:59]: Would suggest a more descriptive name for this function, in particular, useful for what?

File: lib/bib_data_manager.rb [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+        useful = var_fields.inject({ headings: [], published_date: [] }) do |useful_fields, field|
+          next useful_fields unless field['marcTag']
+
+          useful_fields[:headings] << field if field['marcTag'].match %r{^(6\d\d)$}
+          useful_fields[:published_date] << field if PUBLICATION_YEAR_MARC_TAGS.include? field['marcTag']
+
+          useful_fields
+        end
```
Comment (danamansana) [inline lib/bib_data_manager.rb:67]: is `extract_useful_fields` the same as `[var_fields.select {|field| PUBLICATION_YEAR_MARC_TAGS.include? field['marcTag']}, var_fields.select { field['marcTag'].match %r{^(6\d\d)$} }]`

**Review by danamansana — APPROVED** [review]
- Can't really speak much to the correctness since I don't know the use case well enough, will leave that to you
- Suggested a couple of small code style/refactor changes
- Would love to see further reduction of complexity but I also don't want to hold this up.

File: app.rb (hunk 1 above)
Comment (siberry) [reply app.rb:51]: We do not want to call `have_subject_headings_changed?` if `is_research` is false.

Comment (danamansana) [reply app.rb:51]: I think Ruby's `&&` is short-circuiting, so it won't call the second part if the first part is false

Comment (siberry) [reply app.rb:51]: Oh yeah good point

File: lib/bib_data_manager.rb (hunk 3 above)
Comment (siberry) [reply lib/bib_data_manager.rb:59]: This is Stephen's original code

**Pull request merged**